Refactor boxes.py and geometric.py to make them follow the expected convention

mohamedamrfaroukebrahim opened the pull request.

mohamedamrfaroukebrahim committed:
Refactor boxes.py and geometric.py to make them follow the expected convention

mohamedamrfaroukebrahim committed:
Refactor boxes.py and geometric.py to make them follow the expected convention

mohamedamrfaroukebrahim committed:
Refactor boxes.py and geometric.py to make them follow the expected convention

mohamedamrfaroukebrahim committed:
Refactor boxes.py and geometric.py to make them follow the expected convention

Resolve mergingconflict

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
 
 
-def split(boxes):
+def split_into_components(boxes):
```
I think is OK to leave it as "split" only. Since the API that we will be using is paz.boxes.split.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
-    return jp.concatenate(
-        [coordinate_0, coordinate_1, coordinate_2, coordinate_3], axis=1
-    )
+def get_center_coordinates(x_minimum, x_maximum, y_minimum, y_maximum):
```
We can most likely here call it "compute_centers". Since we are always expected to work with "coordinates" and we are "computing" them not "getting them". Moreover, we can here us x_min, y_min, x_max, y_max. We can use "max" and "min" as abbreviations.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
+    return center_x, center_y
+
+
+def get_width_and_height(x_minimum, x_maximum, y_minimum, y_maximum):
```
Here we can call this "compute_size" or "compute_HW". Also we should return H first and then W. The min, max applies here as well.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
+    return predictions[:, 1:2] * priors[:, 3:4] * variances[1] + priors[:, 1:2]
+
+
+def decode_width(predictions, priors, variances):
```
Here we can call it decode_W

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
+    return priors[:, 2:3] * jp.exp(predictions[:, 2:3] * variances[2])
+
+
+def decode_height(predictions, priors, variances):
```
Here can call it decode_H

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
-    return jp.array(boxes)
+
+
+def convert_to_center_form(matched_boxes):
+    """Convert matched boxes to center form."""
+    return to_center_form(matched_boxes[:, :4])
```
I think the outer function does not convey much more meaning that the inner function. Most likely we should call only the inner function when needed.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
-    y_min = center_y - (H / 2.0)
-    y_max = center_y + (H / 2.0)
-    return jp.concatenate([x_min, y_min, x_max, y_max], axis=1)
+    center_x, center_y, W, H = split_into_components(boxes)
```
we should call it "split"

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
+    paz.draw.circle(image, (x, y), radius, color)
+    points.append((x, y))
 
-    # Arguments
-        boxes: Numpy array of shape `(num_boxes, 4)`.
 
-    # Returns
-        Numpy array of shape `(num_boxes, 4)`.
-    """
-    x_min, y_min, x_max, y_max = split(boxes)
-    return join(x_max, y_min, x_min, y_max)
+def process_pair_of_points(image, points, boxes, color, radius):
+    """Process two points to form a box and draw it."""
+    point_A, point_B = take_last_two_points(points)
+    box = order_xyxy(point_A, point_B)
+    paz.draw.box(image, box, color, radius)
+    boxes.append(box)
 
 
-def from_selection(image, radius=5, color=(255, 0, 0), window_name="image"):
-    points, boxes = [], []
-    image = image.copy()
+def on_double_click_event(event, x, y, flags, param, image, points, boxes, radius, color):
+    """Handle double-click event to draw points and boxes."""
+    if event == cv2.EVENT_LBUTTONDBLCLK:
+        draw_circle_and_store_point(image, points, x, y, radius, color)
+        if len(points) % 2 == 0:
+            process_pair_of_points(image, points, boxes, color, radius)
```
This functions will most likely not be needed in the public API. Let's define inside the original function "from_selection" all the functions that are explicitly useful for this case. If we ever see that we might need them somewhere else we can put them out.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -4,377 +4,552 @@
+    """Process final NMS results to get selected indices."""
+    count, _, indices = final_state
+    num_selected = jp.minimum(count, top_k)
+    selected = lax.dynamic_slice(indices, (0,), (num_selected,))
```
I don't know if this function is jittable? Would it be possible to confirm this, maybe a unit test?

oarriaga commented on paz/processors/geometric.py:
```diff
@@ -1,294 +1,505 @@
-    )
-    masked_boxes = masked_boxes.at[:, 2:].set(
-        jp.minimum(masked_boxes[:, 2:], image_crop_box[2:]) - image_crop_box[:2]
+    mask = (
+        (boxes[:, 0] < crop_region[2])
+        & (boxes[:, 1] < crop_region[3])
+        & (boxes[:, 2] > crop_region[0])
+        & (boxes[:, 3] > crop_region[1])
```
This is not so clear to me. I can't tell what coordinates we are comparing. Maybe we can split them before, give them name and then compare them.

mohamedamrfaroukebrahim committed:
Refactor paz/backend/boxes.py and paz/processors/geometric.py with API improvements and JAX optimizations

boxes.py changes:
- Rename functions for consistency:
  * split_into_components → split
  * get_center_coordinates → compute_centers
  * get_width_and_height → compute_size (returns H,W)
  * decode_width → decode_W
  * decode_height → decode_H

- Standardize parameter names:
  * x_minimum → x_min
  * y_minimum → y_min
  * x_maximum → x_max
  * y_maximum → y_max

- Simplify box selection API by Moving helper functions inside from_selection()

- Improve NMS JIT compatibility:

- Remove redundant convert_to_center_form wrapper

geometric.py changes:
- Clarify crop region comparison by:
  * Extracting crop coordinates with descriptive names
  * Splitting box coordinates before comparison
  * Adding explanatory comments for overlap logic

boxes_test.py changes:
- adding a unit test to Test the jitability of process_final_results

mohamedamrfaroukebrahim committed:
Refactor paz/backend/boxes.py and paz/processors/geometric.py with API improvements and JAX optimizations

boxes.py changes:
- Rename functions for consistency:
  * split_into_components → split
  * get_center_coordinates → compute_centers
  * get_width_and_height → compute_size (returns H,W)
  * decode_width → decode_W
  * decode_height → decode_H

- Standardize parameter names:
  * x_minimum → x_min
  * y_minimum → y_min
  * x_maximum → x_max
  * y_maximum → y_max

- Simplify box selection API by Moving helper functions inside from_selection()

- Improve NMS JIT compatibility:

- Remove redundant convert_to_center_form wrapper

geometric.py changes:
- Clarify crop region comparison by:
  * Extracting crop coordinates with descriptive names
  * Splitting box coordinates before comparison
  * Adding explanatory comments for overlap logic

boxes_test.py changes:
- adding a unit test to Test the jitability of process_final_results

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -5,60 +5,47 @@
-        Numpy array with shape `(num_boxes, 4)`.
-    """
-    x_min, y_min, x_max, y_max = split(boxes)
+def compute_centers(x_min, x_max, y_min, y_max):
```
I think this function should take boxes as input, not the coordinates given that we will be calling:
`paz.boxes.computer_center(boxes)`

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -5,60 +5,47 @@
-def xyxy_to_xywh(boxes):
-    x_min, y_min = boxes[:, 0:1], boxes[:, 1:2]
-    x_max, y_max = boxes[:, 2:3], boxes[:, 3:4]
+def compute_size(x_min, x_max, y_min, y_max):
```
This function should take as input boxes not the coordinates. If you are using this function as coordinates, consider splitting the boxes first the applying this function. If you are using this function and the coordinates are scalar values. Then  , let's define the function inside the function that is using it.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -150,231 +120,434 @@ def key_is_pressed(key="q", time=20):
+    return inner_w * inner_h
+
+
+def calculate_union(
```
Most  of the time we say "compute" rather than "calculate" since compute is smaller.

oarriaga commented on paz/backend/boxes.py:
```diff
@@ -150,231 +120,434 @@ def key_is_pressed(key="q", time=20):
+    return inner_x_min, inner_y_min, inner_x_max, inner_y_max
+
+
+def calculate_intersection(inner_x_min, inner_y_min, inner_x_max, inner_y_max):
```
Most of the time we say "compute" rather than "calculate" since compute is smaller.

oarriaga reviewed (changes requested):
There are some minor changes.

mohamedamrfaroukebrahim committed:
Refactor paz/backend/boxes.py and paz/processors/geometric.py with API improvements and JAX optimizations

boxes.py changes:

- Rename functions for consistency:
 * calculate_intersection -> compute_intersection
 * calculate_union -> compute_union

- update functions signature to accept the entire array of boxes:
  * compute_size
  * compute_centers

mohamedamrfaroukebrahim committed:
Merge branch 'paz-jax' of https://github.com/mohamedamrfaroukebrahim/paz into paz-jax

oarriaga committed:
Merge branch 'paz-jax' into paz-jax
